x-pack/solutions/observability/plugins/observability_agent_builder/server/routes/ai_insights/apm_error/fetch_trace_context.ts

```diff
@@ -0,0 +1,230 @@
+  errorExceptionHandled?: boolean;
+  errorExceptionType?: string;
+}
+
+interface ServiceAggregate {
+  serviceName: string;
+  count: number;
+  errorCount: number;
+}
+
+export interface TraceContext {
+  traceItems: TraceItem[];
+  traceServiceAggregates: ServiceAggregate[];
+  traceErrors: TraceError[];
+}
+
+export async function fetchTraceContext({
+  esClient,
+  apmIndices,
+  traceId,
+  start,
+  end,
+  logger,
+}: {
+  esClient: IScopedClusterClient;
+  apmIndices: APMIndices;
+  traceId: string;
+  start: number;
+  end: number;
+  logger: Logger;
+}): Promise<TraceContext> {
+  const search = getTypedSearch(esClient.asCurrentUser);
+  const traceIndices = [apmIndices.transaction, apmIndices.span].join(',');
+  const errorIndices = apmIndices.error;
+
+  const timeRangeFilter = { range: { '@timestamp': { gte: start, lte: end } } };
+  const traceFilter = { term: { 'trace.id': traceId } };
+
+  // Fetch trace documents (transactions and spans)
+  const traceResponse = await search({
+    index: traceIndices,
+    size: 100,
+    track_total_hits: false,
+    query: { bool: { filter: [traceFilter, timeRangeFilter] } },
+    fields: [
+      '@timestamp',
+      'service.name',
+      'trace.id',
+      'event.outcome',
+      'parent.id',
+      'processor.event',
+      'transaction.id',
+      'transaction.name',
+      'transaction.type',
+      'transaction.duration.us',
+      'span.id',
+      'span.name',
+      'span.type',
+      'span.subtype',
+      'span.duration.us',
+      'span.destination.service.resource',
+    ],
+    _source: false,
+    sort: [{ '@timestamp': 'asc' }],
+  });
+
+  const traceItems: TraceItem[] = traceResponse.hits.hits.map((hit) => {
+    const fields = (hit.fields || {}) as Record<string, any[]>;
+
+    const processorEvent = fields['processor.event']?.[0];
+    const isTransaction = processorEvent === 'transaction';
+
+    const baseItem = {
+      timestamp: fields['@timestamp']?.[0],
+      serviceName: fields['service.name']?.[0],
+      traceId: fields['trace.id']?.[0],
+      eventOutcome: fields['event.outcome']?.[0],
+      parentId: fields['parent.id']?.[0],
+    };
+
+    // for transactions
+    if (isTransaction) {
+      return {
+        ...baseItem,
+        transactionId: fields['transaction.id']?.[0],
+        transactionName: fields['transaction.name']?.[0],
+        transactionType: fields['transaction.type']?.[0],
+        transactionDurationUs: fields['transaction.duration.us']?.[0],
+        spanId: fields['span.id']?.[0],
+      };
+    }
+
+    // for spans
+    return {
+      ...baseItem,
+      transactionId: fields['transaction.id']?.[0],
+      spanId: fields['span.id']?.[0],
+      spanName: fields['span.name']?.[0],
+      spanType: fields['span.type']?.[0],
+      spanSubtype: fields['span.subtype']?.[0],
+      spanDurationUs: fields['span.duration.us']?.[0],
+      downstreamServiceResource: fields['span.destination.service.resource']?.[0],
+    };
+  });
```

Comment:
I don't understand this custom formatting. Why not return the docs as they are retuned from ES?